### PR TITLE
feat(datasource): support range constraints filtering

### LINKED ISSUE #8476: Support range/range compatibility filtering
**What would you like Renovate to be able to do?**

Support the scenario where both the repo and datasource have a compatibility range. e.g. the repo has python `>=3.7` and the package has `>=3.6`.

**Did you already have any implementation ideas?**

You could try doing it algebraically, or practically. 

Algebraic seems hard.

If it were practical, you'd use an actual datasource to identify all the possible versions (e.g. of python) and then compute matching values for both ranges before checking that the second range is a subset of the first.

### LINKED ISSUE #18715: PHP lookup constraints
### What would you like Renovate to be able to do?

Support php constraints in composer dependency lookups. i.e. do not propose an update if it is incompatible with the current `php` constraint.

For example, with these constraints: https://github.com/laminas/laminas-diagnostics/blob/df6b5e8d0874f7317b834901c1f6a137ce42ffee/composer.json#L20-L24

Do not propose this change: https://github.com/laminas/laminas-diagnostics/pull/54/files

Because (copy pasted from `composer` output):

```
Root composer.json requires laminas/laminas-loader ^2.9.0 -> satisfiable by laminas/laminas-loader[2.9.0].
    - laminas/laminas-loader 2.9.0 requires php ~8.0.0 || ~8.1.0 || ~8.2.0 -> your php version (7.4.99; overridden via config.platform, actual: 7.4.32) does not satisfy that requirement.
```

Ref: https://packagist.org/packages/laminas/laminas-loader
Ref: https://github.com/laminas/laminas-loader/blob/51ed9c3fa42d1098a9997571730c0cbf42d078d3/composer.json#L24-L26

Note: this seems different to how libraries in Node.js do it, because in Node if you removed support for a major version of the language (e.g. PHP 7) then you should issue a breaking change of your library. The use of e.g. `7.4.99` is also a bit of a composer quirk too.

### If you have any ideas on how this should be implemented, please tell us here.

Related discussion with @Ocramius and @internalsystemerror here: https://github.com/laminas/.github/issues/41

So here we have the case where the downstream repo supports `7.4.99` while the new version of the lib requires `~8.0.0 || ~8.1.0 || ~8.2.0`.

Obviously to the human eye these are incompatible, but we need a foolproof way for the bot to determine it, because the "Venn diagrams" will not always be so clear. For example if the downstream repo supported `7.4.99 || 8.0.99` then would this be OK or not OK? (in my opinion: not ok).

Would this work?

- Find all versions which the downstream repo supports (in this simplified case it's only `["7.4.99", "8.0.99"]`
- Find all versions which the upgrading lib version supports (in this case it's essentially every version of 8.x)
- Reject the upgrade if the second set does not contain every member of the first set. i.e. the lib must support a superset of the downstream consumer in order to be compatible

BTW this could unfortunately let us end up in "stalemates" quite easily where no dependency upgrades are proposed due to the downstream repo having too old or too narrow php requirements, something we should be aware of.

Ideally needs a simplified reproduction because the referenced repo seems quite complicated and therefore difficult to debug.

### Is this a feature you are interested in implementing yourself?

Maybe

### PR DESCRIPTION
**Pull request by Lctrs**
## Changes

- Adds an optional method subset to VersioningApi interface that returns true if first argument is a subset of second argument, false otherwise
- Adds implementations for it in npm, composer, and python versioning
- Uses it in releases filtering in strict constraints mode with fallback to matching

## Context

- Revival of #20217 following branch `v35` merge
- Closes #8476
- Closes #18715
- Closes #19851 

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by rarkins — CHANGES_REQUESTED** [review]
Needs coverage update

**Comment by rarkins** [comment]
Missing coverage:

image

**Review by viceice — DISMISSED** [review]

**Comment by rarkins** [comment]
@Lctrs have you done any recent sanity test on a real repo?

**Comment by Lctrs** [comment]
@rarkins just did it in https://github.com/Lctrs/playground/pull/2. Still works well.

File: docs/usage/configuration-options.md [hunk 1]
```diff
@@ -555,7 +555,7 @@ Renovate supports two options:
 We are working on adding more advanced filtering options.
 
 Note: There must be a `constraints` object in your Renovate config for this to work.
-This feature is limited to `pypi` datasource only.
+This feature is limited to `composer`, `npm`, `pypi` and `poetry` datasources.
```
Comment (rarkins) [inline docs/usage/configuration-options.md:558]:
```suggestion
This feature is limited to `packagist`, `npm`, and `pypi` datasources.
```

**Review by rarkins — CHANGES_REQUESTED** [review]

**Review by rarkins — APPROVED** [review]

**Pull request merged**